### PR TITLE
Add PNN architectural strategy

### PR DESCRIPTION
**Pull request by nmcnamee24**
Adds a compact Progressive Neural Network strategy under `architectural/pnn.py`.

The strategy creates one frozen model column per concept, tracks concept-to-task routing, and supports task-aware prediction as well as optional task-free prediction by selecting the lowest anomaly score across columns.

### CODE REVIEW

File: src/pyclad/strategies/architectural/pnn.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    @property
+    def current_task(self) -> int:
+        return self._current_task
+
+    @property
+    def num_columns(self) -> int:
+        return len(self._columns)
+
+    def learn(self, data: np.ndarray, concept_id: str | None = None, **kwargs) -> None:
+        del kwargs
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py]: The same issue as in earlier PRs. `del kwargs` is not needed and is usually AI leftover

File: src/pyclad/strategies/architectural/pnn.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+import numpy as np
+import torch
+from torch.utils.data import DataLoader, TensorDataset
+
+from pyclad.models.torch_backbone import TorchBackbone
+from pyclad.strategies.strategy import ConceptAwareStrategy, ConceptIncrementalStrategy
+
+
+class PNNStrategy(ConceptIncrementalStrategy, ConceptAwareStrategy):
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:8]: Do we need the idea of `task_label` here? Won't `concept_id` be sufficient?

File: src/pyclad/strategies/architectural/pnn.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        if not self._columns:
+            raise ValueError("PNNStrategy cannot predict before learning at least one concept.")
+
+        if task_label is not None:
+            return self._column(task_label).predict(data)
+        if concept_id is not None and concept_id in self._concept_to_task:
+            return self._column(self._concept_to_task[concept_id]).predict(data)
+        if self._task_free:
+            return self._task_free_predict(data)
+        return self._column(self._current_task).predict(data)
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py]: I don't think we should have default behavior here. If `concept_id is None` and `not self._task_free` it seems like a mistake. Right now, if we run PNN in concept incremental scenario with `task_free=False`, we would get very strange results, as it would always use column for current task

File: src/pyclad/strategies/architectural/pnn.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    @property
+    def current_task(self) -> int:
+        return self._current_task
+
+    @property
+    def num_columns(self) -> int:
+        return len(self._columns)
+
+    def learn(self, data: np.ndarray, concept_id: str | None = None, **kwargs) -> None:
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:16]: Should we allow learning the same concept twice? or should we have a safeguard raising an error in such case?

File: src/pyclad/strategies/architectural/pnn.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            self._concept_to_task[concept_id] = task_label
+        return task_label
+
+    def _fit_column(self, task_label: int, data: np.ndarray) -> None:
+        model = self._column(task_label)
+        self._set_trainable(model, True)
+        current = np.asarray(data, dtype=np.float32)
+        loader = DataLoader(TensorDataset(torch.tensor(current, dtype=torch.float32)), self._batch_size, shuffle=True)
+
+        def loss_fn(batch) -> torch.Tensor:
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py]: Couldn't we do: 
`model.fit_with_loss(loader, model.compute_loss, ...` 
and avoid defining `loss_fn` entirely?

File: tests/strategies/smoke_tests/test_pnn.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from copy import deepcopy
+
+import pytest
+
+from pyclad.strategies.architectural.pnn import PNNStrategy
+from tests.strategies.smoke_tests.base import BaseStrategyTest
+
+
+class TestPNNStrategy(BaseStrategyTest):
```
Comment (Nyderx) [inline tests/strategies/smoke_tests/test_pnn.py:142]: We should also cover the case with `task_free=False`

File: src/pyclad/strategies/architectural/pnn.py
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:1]: One issue about task_free approach:

>   1. Task-free prediction selects argmin of uncalibrated scores — likely wrong (pnn.py:93–100)
> 
>   best = np.argmin(scores, axis=0)
>   return labels[best, sample_index], scores[best, sample_index]
> 
>   Selecting the column with the lowest reconstruction error per sample assumes scores are comparable across columns. They're not — each column trains independently with its own loss scale. A column trained on compact, low-variance data will systematically produce smaller errors than one trained on a wider
>   distribution, so argmin will almost always select the same column regardless of which concept the sample actually came from.
> 
>   The returned score is then that minimum — which will be biased low, blunting anomaly detection sensitivity. The more principled approach is to normalize each column's scores (e.g. z-score relative to that column's training loss distribution) before aggregating, or to use the column that was trained on the
>    most similar distribution as identified by some task-ID estimator.

It's something to discuss with Roberto as well

**Review by Nyderx — CHANGES_REQUESTED** [review]

File: src/pyclad/strategies/architectural/pnn.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        **kwargs,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        if not self._columns:
+            raise ValueError("PNNStrategy cannot predict before learning at least one concept.")
+
+        if concept_id is not None:
+            if concept_id not in self._concept_to_column:
+                raise ValueError(f"Unknown concept_id '{concept_id}'.")
+            return self._column(self._concept_to_column[concept_id]).predict(data)
+        if self._task_free:
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:34]: Lines 109-110 should be before 105 to properly handle the situation in which we have `self._task_free = True` and someone passes `concept_id`.

File: src/pyclad/strategies/architectural/pnn.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        self._epochs = epochs
+        self._task_free = task_free
+        self._device = torch.device(device)
+        self._columns: list[TorchBackbone] = []
+        self._column_score_stats: list[tuple[float, float]] = []
+        self._concept_to_column: dict[str, int] = {}
+        self._current_column = -1
+
+    @property
+    def current_column(self) -> int:
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py]: Why do we need `self._current_column` at all?

File: src/pyclad/strategies/architectural/pnn.py
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:1]: I am not sure about something.

PNN is supposed to have lateral connections between columns, right? each new column receives activations from all previously frozen columns as additional inputs.

Do we have lateral connections between columns in this implementation? 

If every column is a completely independent, identically initialised backbone trained in isolation, it seem to be  equivalent to "one model per concept" (like MSTE).

File: src/pyclad/strategies/architectural/pnn.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        return self._predict_column(self._concept_to_column[concept_id], data) if concept_id in self._concept_to_column else (np.zeros(data.shape[0]), np.zeros(data.shape[0]))
+    def _fit(self, column_index: int, data: np.ndarray) -> None:
+        loader = DataLoader(torch.tensor(data, dtype=torch.float32, device=self._device), self._batch_size, shuffle=True)
+        optimizer = None; self._columns[column_index].get_module().train(); self._adapters[column_index].train()
+        for _ in range(self._epochs):
+            for batch in loader:
+                loss = self._loss(column_index, batch); optimizer = optimizer or self._optimizer(column_index)
+                optimizer.zero_grad(); loss.backward(); optimizer.step()
+    def _loss(self, column_index: int, x: torch.Tensor) -> torch.Tensor:
+        return self._columns[column_index].compute_loss(x) if column_index == 0 else F.mse_loss(self._forward(column_index, x), x)
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:47]: **Bug: MSE hardcoded as training loss for all columns after the first.**

`compute_loss` is used only when `column_index == 0`; every subsequent column trains under `F.mse_loss` regardless of what the backbone declares. A VAE (ELBO loss), a model with BCE, or any backbone with a custom objective will silently train under the wrong loss for columns 1+. Suggest using `self._columns[column_index].compute_loss(self._forward(column_index, x))` for all columns, or document the MSE-only restriction explicitly.

File: src/pyclad/strategies/architectural/pnn.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        best = np.expand_dims(np.argmin(scores, axis=0), axis=0)
+        return np.take_along_axis(labels, best, axis=0).squeeze(axis=0), np.take_along_axis(scores, best, axis=0).squeeze(axis=0)
+    def _predict_column(self, column_index: int, data: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        if column_index == 0:
+            return self._columns[column_index].predict(data)
+        data = np.asarray(data, dtype=np.float32)
+        with torch.no_grad():
+            x_hat = self._forward(column_index, torch.tensor(data, dtype=torch.float32, device=self._device)).cpu().numpy()
+        scores = ((data - x_hat) ** 2).mean(axis=1)
+        return (scores > getattr(self._columns[column_index], "threshold", 0.5)).astype(int), scores
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:75]: **Bug: Column 0 and columns 1+ use incommensurable prediction paths.**

Column 0 returns whatever `backbone.predict()` produces (any score scale, any thresholding). Columns 1+ always compute MSE reconstruction error and threshold with `getattr(backbone, "threshold", 0.5)`. In task-free mode, `_predict_task_free` z-normalises and compares scores across all columns. If the backbone for column 0 uses a different scoring scale (log-likelihoods, bounded probabilities), z-normalisation cannot align them with MSE-based scores, and the column selector will systematically misroute. All columns should go through the same prediction path.

File: src/pyclad/strategies/architectural/pnn.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        self._set_trainable(column_index, True); self._fit(column_index, data); self._set_trainable(column_index, False)
+        self._score_stats.append(self._score_stats_for(column_index, data))
+    def predict(self, data: np.ndarray, concept_id: str | None = None, **kwargs) -> tuple[np.ndarray, np.ndarray]:
+        if not self._columns:
+            raise ValueError("PNNStrategy cannot predict before learning at least one concept.")
+        if self._task_free:
+            return self._predict_task_free(data)
+        if concept_id is None:
+            raise ValueError("PNNStrategy requires concept_id for prediction unless task_free=True.")
+        return self._predict_column(self._concept_to_column[concept_id], data) if concept_id in self._concept_to_column else (np.zeros(data.shape[0]), np.zeros(data.shape[0]))
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:38]: **Bug: Unknown `concept_id` silently returns all-zero scores.**

When `concept_id` is not in `_concept_to_column`, `predict` returns `(np.zeros(n), np.zeros(n))` instead of raising. Score 0.0 is below any reasonable anomaly threshold, so a typo or misconfigured concept ID causes every sample to be silently classified as normal. This is a regression from the previous design, which raised `ValueError`. Suggest raising `ValueError(f"Unknown concept_id '{concept_id}'.")` here instead.

File: src/pyclad/strategies/architectural/pnn.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        if column_index == 0:
+            return self._columns[column_index].predict(data)
+        data = np.asarray(data, dtype=np.float32)
+        with torch.no_grad():
+            x_hat = self._forward(column_index, torch.tensor(data, dtype=torch.float32, device=self._device)).cpu().numpy()
+        scores = ((data - x_hat) ** 2).mean(axis=1)
+        return (scores > getattr(self._columns[column_index], "threshold", 0.5)).astype(int), scores
+    def _layers(self, column_index: int) -> list[nn.Module]:
+        module = self._columns[column_index].get_module(); layers = []
+        for block in (module.encoder, module.decoder):
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:78]: **Altitude: `_layers()` reaches past `TorchBackbone` into raw module attributes.**

Accessing `module.encoder` and `module.decoder` directly couples the strategy to the backbone's internal layout. Any backbone that stores its layers under different names (`self.blocks`, `self.net`) fails at `learn()` time with a `ValueError`, even if it is otherwise fully conformant with `TorchBackbone`. The right fix is to add `layers() -> list[nn.Module]` as an abstract method on `TorchBackbone` (with each backbone declaring its own layer sequence), removing the attribute sniffing entirely.

File: src/pyclad/strategies/architectural/pnn.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        h = x
+        for layer_index, layer in enumerate(self._layers(column_index)[: stop + 1 if stop is not None else None]):
+            h = layer(h)
+            if column_index:
+                with torch.no_grad():
+                    old = [self._forward(index, x, layer_index, cache).detach() for index in range(column_index)]
+                adapters = self._adapters[column_index][layer_index]
+                while len(adapters) < len(old):
+                    adapter = nn.Linear(old[len(adapters)].flatten(1).shape[1], h.flatten(1).shape[1], bias=False).to(h.device)
+                    adapter.requires_grad_(adapters.training); adapters.append(adapter)
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:93]: **Bug: Lazy adapter `requires_grad` inferred from `adapters.training`, which may already be `False`.**

`adapter.requires_grad_(adapters.training)` sets the gradient flag from the parent `ModuleList`'s training flag at the moment of creation. Adapters are created lazily inside `_forward` during `_fit` — but if `_set_trainable` ordering ever changes or a previous `learn()` was interrupted, `adapters.training` could be `False` at creation time, silently producing permanently-frozen adapters with no error. Use `adapter.requires_grad_(True)` unconditionally here; the adapter is always being created during a training pass.

File: src/pyclad/strategies/architectural/pnn.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    def _params(self, column_index: int):
+        yield from self._columns[column_index].get_module().parameters()
+        yield from self._adapters[column_index].parameters()
+    def _set_trainable(self, column_index: int, trainable: bool) -> None:
+        self._columns[column_index].get_module().train(mode=trainable); self._adapters[column_index].train(mode=trainable)
+        for parameter in self._params(column_index):
+            parameter.requires_grad = trainable
+    def _score_stats_for(self, column_index: int, data: np.ndarray) -> tuple[float, float]:
+        _, scores = self._predict_column(column_index, np.asarray(data, dtype=np.float32)); scores = np.asarray(scores, dtype=np.float32)
+        return float(scores.mean()), float(scores.std()) or 1.0
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:60]: **Bug: `float(scores.std()) or 1.0` does not guard NaN.**

The intent is to fall back to `1.0` when std is zero. `float(0.0)` is falsy, so that case works. But `float(nan)` is truthy in Python, so `float(nan) or 1.0` returns `nan` unchanged. If any backbone produces NaN anomaly scores during training, NaN propagates into the normalisation denominator in `_predict_task_free`, silently corrupting all task-free predictions. Use `s if (s := float(scores.std())) > 0.0 else 1.0` — NaN comparisons return `False`, making this guard NaN-safe as well.

File: src/pyclad/strategies/architectural/pnn.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
+from typing import Callable
+import numpy as np, torch
```
Comment (Nyderx) [inline src/pyclad/strategies/architectural/pnn.py:3]: **Convention: Multiple statements per line via semicolons fail project linting.**

The file uses semicolons throughout (e.g. line 5: `from torch import nn; from torch.utils.data import DataLoader`, line 29: `self._set_trainable(...); self._fit(...); self._set_trainable(...)`, line 93: `adapter.requires_grad_(...); adapters.append(adapter)`). CONTRIBUTING.md requires `black src` and `flake8 src`; flake8 E702 fires on every semicolon-separated statement and E401 fires on the comma-joined import on this line. Every other file in the project uses standard single-statement-per-line formatting.

**Review by Nyderx — COMMENTED** [review]
Review of the updated PNN implementation. Eight findings below — five are correctness issues that affect any backbone other than `Autoencoder`.
